### DISCUSSION TITLE
how to check if nerd font is installed

### DISCUSSION THREAD

**Discussion by Araxeus** [ryanoasis/nerd-fonts#829] [Q&A]
I want to implement nerd font in my application (https://github.com/Araxeus/ls-interactive)

But I want it to automatically detect if nerd font is enabled in the current terminal (meaning the icons will render properly)
So that it can fallback to Emoji / text instead

Is there any way of doing this?

**Comment by Finii** [comment]
I read you description as this:
* You have a command line application
* That runs in any (unspecified) terminal emulator / console
* You want to find out if a specific codepoint is rendered or unknown

TL;DR: _no_

I guess that is not possible generally. Usually you talk to the console through ANSI sequences, but I guess that there is no function that you would need. You can query the current background color for example to automatically use dark or light 'skin' colors; but glyph rendering specifics? I doubt that.

If you find out in which terminal your application is executed in (which is possible) you might be able to talk to it via signals or something, but that would be very terminal specific. Or maybe some terminals allow to be queried (by executing them) some font details, but that also would be terminal specific.

What I want to say, I believe there is no general way to do this. There might be a terminal specific way to do it, but I deem even that unlikely.
You would at least need to find out which font is the main used font. Then you could check yourself with system tools. That would ignore system wide font fallback (if supported). And then there are terminal emulators that support multiple active fonts for different codepoint ranges.

**Reply by Araxeus** [reply]
Thank you for the detailed answer.

I suspected it wasn't possible but asked anyways in case i was wrong :P

Maybe somehow render an icon to terminal and read the output, if the output is � then it means it isn't supported?

But I don't think you can actually read the line output in terminal right?

It kinda sucks that it is this way, because if I can't check that a user has a valid font, I just won't include special Icons :/

**Reply by Finii** [reply]
Even _if_ it is rendered, it might be something you do not expect.
Like people who use the Apple glyph at `F8FF` - where Apple and only Apple put it - and complain it has a different location in other (non Apple) fonts (#492) and their complete stash of documents become 'corrupted'.

> render an icon to terminal

Although that is indeed possible to somehow come up with a solution, but that would be highly unportable. For me that sounds like something I do for tests in a tightly controlled container.
Maybe better ask this at some terminal emulator repos.

_Edit: Fix issue link_

**Reply by bschlenk** [reply]
I was wondering this as well, and came to the same conclusion.

Should there be a convention of using a specific env variable to mean "I am using a nerd font"? Like `export NERD_FONT=1`? That way you could export a single env variable and expect any cli app that uses nerd fonts to work correctly in your terminal.

**Reply by Finii** [reply]
Who should set that variable reliably? Fonts can not set variables.
Some install script could add something via `/etc/profile.d/`, but that in itself is also rather specific. But what if a person uses a different way of installation?
And even _if_ we can set a variable, who guarantees that it is used by the terminal?

You can only put that into your application and instruct your users that they can/have to give `--use-nerd-font-icons` on the command line or something, or set an environment var or another if they want your application to utilize codepoints that might or might not be in a non-Nerd-Font.

I guess this is somehow like 'has the user a color terminal or not?' of old. It was (is?) not really possible to determine that. And `ls` has a switch to turn ansi color on because the automatism is so flaky...

```
NAME
       ls - list directory contents

SYNOPSIS
       ls [OPTION]... [FILE]...

DESCRIPTION
[...]
              --color[=WHEN]
              colorize the output; WHEN can be 'always'  (default
              if omitted), 'auto', or 'never'; more info below
[...]
       Using  color to distinguish file types is disabled both by
       default and with  --color=never.   With  --color=auto,  ls
       emits  color  codes only when standard output is connected
       to a terminal.  The  LS_COLORS  environment  variable  can
       change the settings.  Use the dircolors command to set it.

```

Users that want to control color put an alias into their startup script, like `alias ls=ls --color=never`.

**Reply by evnoj** [reply]
> Should there be a convention of using a specific env variable to mean "I am using a nerd font"? Like `export NERD_FONT=1`? That way you could export a single env variable and expect any cli app that uses nerd fonts to work correctly in your terminal.

I like this. I've written a few scripts and tools that I'd like to use nerd font when it's available, but by default just use ASCII. These tools mostly aren't public (yet), but they only use nerd font symbols if `NERD_FONT` is present. I like this more than passing an option to each tool, although any that I intend for others to use will also support a command-line option.

> Who should set that variable reliably?

The user (or their shell config framework) can set this variable, same as they can pass `--use-nerd-font-icons` to a command. Since this option will likely be shared across tools in a session, an environment variable seems like a fine way of doing it. `ls` (and other tools) supports enabling color based on the presence of the `CLICOLOR` environment variable, in addition to command-line options, for that reason.

I arrived here seeking some convention, but it doesn't seem that there is one, so this is good enough for me.
